Add calculated numeric fields

romseygeek opened the pull request:
This commit adds a `script` parameter to `long` and `double` fields that makes
it possible to calculate a value for these fields at index time.  It uses the same
script context as the equivalent runtime fields, and allows for multiple index-time
scripted feels to cross-refer while still checking for indirection loops.

romseygeek committed:
Take 3... Just re-use scripts, exposing things via a MemoryIndex if necessary

romseygeek committed:
add test for cross references

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -476,6 +508,11 @@ public Short parse(XContentParser parser, boolean coerce) throws IOException {
 
+            @Override
+            public BiFunction<Script, ScriptService, MapperScript> compiler(String fieldName) {
+                return LONG.compiler(fieldName);
```
Is it ok to "silently" promote here? I think I'd prefer a more conservative "no, you can't make this one calculated, use long" message for now and to relax it in the future.

romseygeek replied:
++, have changed these to throw explicit errors

nik9000 reviewed (commented):
Took a look. Don't know how I feel about the memory index yet. I guess I'll have to reread its source code....

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Only support script on long and double

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Allow calculated fields to refer to each other

romseygeek committed:
Add value fetchers; yaml tests

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+    private IndexTimeScriptContext(MappingLookup mappingLookup, ParseContext pc) {
+        this.searchLookup = new SearchLookup(
+            mappingLookup::getFieldType,
+            (ft, s) -> ft.fielddataBuilder("", s).build(
```
This is a TODO - we should be able to get the index name from somewhere and pass it through here.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+    private void executeFieldScript(String field) throws IOException {
+        assert mapperScripts.containsKey(field);
+        mapperScripts.get(field).script.accept(this);
+        mapperScripts.get(field).script = c -> {};
```
This stops scripts being executed multiple times if they are referred to by other scripts

nik9000 replied:
It might be better to set a boolean. I expect there won't be a real speed difference and it'd probably be a bit easier to follow.

romseygeek replied:
++

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Avoid null metadata

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Pass index name; throw error if you try and index into a script field

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+            mappingLookup::getFieldType,
+            (ft, s) -> ft.fielddataBuilder(pc.indexSettings().getIndex().getName(), s).build(
+                new IndexFieldDataCache.None(),
+                new NoneCircuitBreakerService())
```
`NoneCircuitBreakerService` would be bad if this bumped into fielddata. Not having the cache is probably also bad here.

romseygeek replied:
The builder will only be used for a single document, and it won't actually work with fielddata as we don't support `terms()` on the lazy reader, so I don't think either of these will be a problem?

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -55,6 +57,7 @@ private CacheKey() {}
     private final boolean hasNested;
     private final FieldTypeLookup fieldTypeLookup;
     private final Map<String, NamedAnalyzer> indexAnalyzersMap = new HashMap<>();
+    private final Map<String, CheckedConsumer<IndexTimeScriptContext, IOException>> postParsePhases = new HashMap<>();
```
Maybe a name that implies the script context?

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -958,6 +1025,28 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+                    public List<Object> fetchValues(SourceLookup lookup) {
+                        List<Object> values = new ArrayList<>();
+                        script.executeAndIndex(context.lookup(), ctx, lookup.docId(), values::add);
+                        values.sort(Comparator.comparingLong(a -> (Long) a));
```
I believe we're ok with getting the values back in whatever order. We don't, for example, sort the values from source in SearchValueFetcher.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1088,8 +1188,75 @@ protected void parseCreateField(ParseContext context) throws IOException {
+
+        @Override
+        public void executeAndIndex(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter) {
+            LongFieldScript s = scriptFactory
```
Do you think it'd be cool if `LongFieldScript` took a `LongConsumer`?

romseygeek replied:
As in, `runFocDoc(doc, consumer)` type of thing?

nik9000 replied:
It'd be reasonable to save that for a follow up. I'm convinced it'd save code and wouldn't slow things down but I'm not convinced it'd be a hell of a lot cleaner. It's not a blocker at least.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/ScriptParams.java:
```diff
@@ -0,0 +1,76 @@
+                    return null;
+                }
+                Script script = Script.parse(o);
+                return new CompiledScriptParameter<>(script, compiler.apply(script, c.scriptService()));
```
In runtime fields we don't allow indexed scripts because storing them in the mapping would get super confusing. We probably should do that here too.

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+                }
+            }
+            mi.freeze();
+            this.in = mi.createSearcher().getIndexReader().leaves().get(0);
```
Do you think we could build have a `Map` of field name to `LeafReaderContext` and just add the things we need to it? That'd save all the adding if there are dependencies.

nik9000 reviewed (commented):
Could you add a test for when a calculated field depends on a runtime field? I think we said we didn't want that to work, but I think we should send a nice error message back to folks. It'd be *lovely* if we could do that on mapping update, but that may be a "for later" thing.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+        mapperScripts.get(field).script = c -> {};
+    }
+
+    private static class MapperScript {
```
It is really unclear what this class is for. It looks like it holds the post parse phase, hence maybe if that phase becomes its own class, maybe this is no longer needed or at least clearer?

romseygeek replied:
It holds the post parse phase but also ensures that it only gets executed once; if calculated fieldA refers to calculated fieldB, then fieldB's script will get executed as part of fieldA's execution, so it doesn't need to run again on its own.  I'll rename it.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -137,6 +140,10 @@ public MappingLookup(Mapping mapping,
                 throw new MapperParsingException("Field [" + mapper.name() + "] is defined more than once");
             }
             indexAnalyzersMap.putAll(mapper.indexAnalyzers());
+            CheckedConsumer<IndexTimeScriptContext, IOException> postParsePhase = mapper.getPostParsePhase();
```
I am thinking that clarifying and consolidating the notion of post parse phase, or whatever we want to call it, we could even make it more specific to script execution because that is all we need it for, would help making this more readable too.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1043,6 +1134,11 @@ protected String contentType() {
     protected void parseCreateField(ParseContext context) throws IOException {
+
+        if (this.script != null) {
+            throw new IllegalArgumentException("Cannot index data directly into scripted field");
```
meta: we need to come up with a name that we all agree on for this functionality, and use that name consistently

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1088,8 +1188,75 @@ protected void parseCreateField(ParseContext context) throws IOException {
     }
+
+    private interface NumberMapperScript {
+        void executeAndIndex(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter);
```
I find the name a bit distracting because effectively indexing is not what we do in all cases. do we rather collect values and provide them one by one to the emitter consumer?

romseygeek replied:
Yes, `executeAndEmit` is better

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1088,8 +1188,75 @@ protected void parseCreateField(ParseContext context) throws IOException {
+            double[] vs = s.values();
+            for (int i = 0; i < s.count(); i++) {
+                emitter.accept(vs[i]);
+            }
```
can't help but wonder if these two impls could share more of their code. I guess that it all starts with the factories not sharing a base class, and I vaguely remember talking with @nik9000 about this not being straight-forward...

romseygeek replied:
Yes, I tried this, but it's the same problem as with the various runtime field types - if we want to avoid boxing then everything needs to have a separate implementation to handle primitive types.

nik9000 replied:
I think if we shared the `scriptParams` and `fieldName` with a superclass and did the thing I was talking about in the comment above these classes would be very very close to one liners. As close as you can be in java, at any rate.

romseygeek replied:
I've collapsed them both down into anonymous classes returned by their relative numeric types.  Nik's idea will make them even smaller but let's maybe do that later?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1088,8 +1188,75 @@ protected void parseCreateField(ParseContext context) throws IOException {
         return new Builder(simpleName(), type, ignoreMalformedByDefault, coerceByDefault).init(this);
     }
+
+    private interface NumberMapperScript {
```
just by looking at it, it is hard to tell why this class is needed. I think it holds a compiled script, as scripts get compiled when the mappings are parsed. Maybe we should better define what abstractions are needed: it seems that Script is not enough as it is not compiled, and we need some representation for a compiled script, that then gets executed once for each document being indexed. I even wonder if this should be an interface or rather a base class, because there are aspects of a compiled script that all of the types share (params, fieldName)?

romseygeek replied:
We need to make a distinction between a `Script`, which just holds the json representation of a script (so is needed for serialization and comparison during merges, etc); then the compiled script (here either `LongFieldScript` or `DoubleFieldScript`) which is shared with runtime fields; then this mapper script, which executes the compiled script, retrieves its values, and passes it to the indexer.  This interface is the last one, and will probably end up being public and shared - maybe I should look at adding a KeywordFieldMapper implementation here as well, as that should confirm which of these interfaces will be common between different mapper types?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ScriptParams.java:
```diff
@@ -0,0 +1,76 @@
+
+    private ScriptParams() { }
+
+    public static class CompiledScriptParameter<T> implements ToXContent {
```
can you implement ToXContentObject instead?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ScriptParams.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+public final class ScriptParams {
```
I think params are intended for something else in the context of scripts. As far as I understand this class holds the original script as well as the script in its compiled form, as well as the param needed to parse it from the mappings.

romseygeek replied:
Params here are mapper params - we need to hold the original Script for serialization and comparison during merges, but we also only want to compile the script once.

nik9000 replied:
The name of the class is confusing because `Script` has a `getParams` method. This sort of parameter is going to be super common - should it just be a static method inside `FieldMapper.Parameter`? You've done that a few times but not always. I know you had a good reason when you made `TextParams` but I sure don't remember it.

romseygeek replied:
Now that serialization and equality checking is handled directly in MapperScript I think I can just move this parameter definition into FieldMapper.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/ScriptParams.java:
```diff
@@ -0,0 +1,76 @@
+
+    public static class CompiledScriptParameter<T> implements ToXContent {
+        final Script script;
+        final T compiledScript;
```
It is hard to follow what T is here, and I wonder if by any chance this one can be merged with the MapperScript above, or maybe their responsibilities can be clarified. It seems like we need a way to carry around the compiled script, but we do it as part of two different abstractions. Would there be a way to get the compilation out of there, but still have it happen at the right moment?

romseygeek replied:
`T` is in fact the mapper script - javadoc will help here I'm sure

romseygeek commented:
> Could you add a test for when a calculated field depends on a runtime field? I think we said we didn't want that to work, but I think we should send a nice error message back to folks. It'd be lovely if we could do that on mapping update, but that may be a "for later" thing.

At the moment this doesn't throw an error, because we make no distinction in the MappingLookup between runtime fields and concrete fields.  @javanna and I talked about this a bit in #70065 and were unsure if we wanted to make the distinction, but the fact that it's needed in at least two places now (here + index sorting) suggests that we need it.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+    private IndexTimeScriptContext(MappingLookup mappingLookup, ParseContext pc) {
+        this.searchLookup = new SearchLookup(
+            mappingLookup::getFieldType,
+            (ft, s) -> ft.fielddataBuilder(pc.indexSettings().getIndex().getName(), s).build(
```
The index is needed only for IndexFieldMapper, right? Maybe we should rework how it is passed around? Not as part of this change for sure...

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -248,6 +270,13 @@ public Float parse(XContentParser parser, boolean coerce) throws IOException {
             }
 
+            @Override
+            public BiFunction<Script, ScriptService, NumberMapperScript> compiler(String fieldName) {
```
I wish that there were less functions and indirections: it is hard to follow the flow. I guess this has to do with the script service being available only from ParseContext, when mappings are being parsed. And it does need to be on the mapper because the returned type changes for each impl? Is there a chance that we can have an abstraction around a compiled script, with a generic type, that can be used everywhere, and the compiler method does not need to be on every mapper that allows to index calculated values? I see that down the line, a T type is used in CompiledScriptParameter anyways, which throws me off.

romseygeek replied:
Yeah, I think we might be able to combine NumberMapperScript and CompiledScriptParameter, I'll give that a go.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -958,6 +1025,28 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+                if (hasDocValues()) {
+                    return new DocValueFetcher(DocValueFormat.RAW, context.getForField(this));
+                }
+                return new ValueFetcher() {
```
do we have tests for this scenario?

romseygeek replied:
There are specific YAML tests for dv and no-dv cases. Unit testing is trickier but I may be able to make use of some of the bits that Nik plans to add in #70278

javanna reviewed (commented):
Thanks for opening this! I did a first review round and left some design comments, mostly around abstractions and naming. I think that the functionality is there but I am hoping that we can find ways to simplify the flow a bit, I also asked @jtibshirani to have a look.

javanna commented:
> At the moment this doesn't throw an error, because we make no distinction in the MappingLookup between runtime fields and concrete fields. @javanna and I talked about this a bit in #70065 and were unsure if we wanted to make the distinction, but the fact that it's needed in at least two places now (here + index sorting) suggests that we need it.

I don't know the answer yet, but I am wondering if there are ways to detect this without introducing a distinction between runtime fields and concrete fields in `MappingLookup`. I wonder about throwing an error at mapping update also: it sounds like what we should do, but how? Is the plan to find out by executing the script, or to retrieve the fields that the script refers to and check them one by one (and their dependencies, and so on)?

romseygeek commented:
> Is the plan to find out by executing the script, or to retrieve the fields that the script refers to and check them one by one (and their dependencies, and so on)?

I don't think this will be possible, because you can look up fields using an arbitrary string; say you wanted to emit the values from one of a set of fields, based on the value of another field.
```
use_field = doc['use_field'];
emit(doc[use_field]);
```
We can't validate the fields on this without knowing the contents of the document

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/IndexTimeScriptContext.java:
```diff
@@ -0,0 +1,256 @@
+    public final ParseContext pc;
+    private final Map<String, MapperScript> mapperScripts = new HashMap<>();
+
+    public static void executePostParsePhases(MappingLookup lookup, ParseContext parseContext) throws IOException {
```
This is tricky because there's some self-reference going on here with the MapperScripts, etc.  I'm going to try rename things and add javadoc to clarify how it all connects.

romseygeek committed:
Rename, consolidate, add javadocs

javanna commented:
> We can't validate the fields on this without knowing the contents of the document

I vaguely recall that this was discussed before, and the condition to be able to expose fields that a script accesses was that dynamic names were disallowed, maybe @nik9000 remembers more details around this

romseygeek committed:
split postparsecontext into context and phase executor

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Make defining test scripts easier

romseygeek committed:
loop detection

romseygeek committed:
disallow stored scripts

romseygeek committed:
Don't use memory index

romseygeek committed:
tidy

romseygeek commented:
I've iterated further on this:
* no more MemoryIndex, we now build DV impls directly by looking at fields in the lucene Document
* MapperScript is now top level and serialization is directly handled by the ScriptParam
* PostParseContext is just a state holder, execution is now handled by a new PostParsePhase class
* Renamed a bunch of stuff to (hopefully!) make it clearer how everything is related

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:
```diff
@@ -569,4 +569,5 @@ void toXContent(XContentBuilder builder, Params params, ToXContent custom) throw
     protected void doXContent(XContentBuilder builder, Params params) throws IOException {
 
     }
+
```
Maybe remove this from the PR just so I don't have to keep scrolling by it....

romseygeek replied:
die, random newline, die!

nik9000 commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,484 @@
+        false
+    );
+
+    private static NumericDocValues numericDocValues(List<Number> values) {
```
I didn't really like the `MemoryIndex`, but one of the advantages was that it'd enforce the same sort of rounding as you'd get if you loaded from doc values. I *think* this keeps that rounding, but I imagine it'll be a thing we'd have to have a clutch of tests for. If we don't then sneaky things'll happen, I think.

romseygeek replied:
Everything here is a long - `NumericDocValuesField` takes longs, and `NumericDocValues` returns them, so all the rounding and double encoding happens at the mapper level.  I'm going to add some more comprehensive tests directly on the lazy reader which should reassure everybody :)

nik9000 replied:
Say your `long` references a `half_float` field. One we'd build doc values for. If your long were a runtime field it'd get the `half_float` precision. But if its a calculated field will it get `double` precision?

romseygeek replied:
The `half_float` field will have added its values to the underlying `Document` by calling `fields.add(new SortedNumericDocValuesField(name, HalfFloatPoint.halfFloatToSortableShort(value.floatValue())));` (see https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java#L206) which encodes things into a short value, but then casts them to a long for docvalues.  Then when we retrieve the doc values in fielddata builder, things are passed through `SortedNumericIndexFieldData#SingleHalfFloatValues` which reads longs from the underlying NumericDocValues and decodes them back to floats via `HalfFloatPoint.sortableShortToHalfFloat((short) in.longValue());`. Precision doesn't matter for doc values, everything is always stored as `long`.

nik9000 replied:
Those are the sorts of tests that I was worried about. Making sure we lose precision in the same way. It looks like we do *now* but one day someone is going to break it if we don't have a test for it.

romseygeek replied:
Yeah I want to add some tests to MapperTestCase that check for every field type that you can retrieve these values at index time and a search time and they behave the same way, but I think we should leave that to a follow up because it will blow up this PR a lot if we try and add it here.

romseygeek replied:
Also: the MemoryIndex just stores things as longs as well, so doing this via MI or via these hand-rolled DV implementations doesn't make any difference to what we need to test.

nik9000 replied:
:+1: on MemoryIndex.

I don't mind blowing up the PR if it proves that it can handle sneaky stuff. Maybe just a single test case for something sneaky like `half_float` or `float`.

romseygeek committed:
cleanup

romseygeek committed:
better error message

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Throw error if we try to refer to runtime fields in indextime scripts

romseygeek commented on server/src/test/java/org/elasticsearch/index/mapper/CalculatedFieldTests.java:
```diff
@@ -0,0 +1,306 @@
+        }
+    }
+
+    private static class TestDoubleFieldScript extends DoubleFieldScript {
```
These classes make it possible to separate out the `getLongScript` and `getDoubleScript` methods above - they can't take `LongFieldScript` etc directly because `emit` and `extractFromSource` are protected methods.  Another option would be to just make them public?

romseygeek commented:
I've tidied up a bit more and added checks for runtime fields
* ScriptParams is gone
* The repeated code in LongMapperScript and DoubleMapperScript is now subsumed into a couple of much smaller anonymous classes

I want to add some more testing, specifically to check that all mapped field types with docvalues are available via index time scripts, but I think that might be better done in an immediate follow-up?

romseygeek committed:
Add fetch tests for long/double index-time script fields with no docvalues

romseygeek committed:
javadocs

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Add test to MapperTestCase that checks fielddata returned via LazyDocumentReader

romseygeek committed:
Allow test case to say if field does not support scripts

romseygeek committed:
unsigned long doesn't support scripts

romseygeek committed:
or geoshapewithdocvalues

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -958,6 +1051,27 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+            if (this.script != null) {
+                // values don't live in source - either pull from dv, or re-calculate
+                if (hasDocValues()) {
+                    return new DocValueFetcher(DocValueFormat.RAW, context.getForField(this));
```
It'd be good to discuss the upsides/ downsides of this approach. We've warned users against loading too many doc value fields, since each one involves a seek. We enforce a soft limit of 100 doc value fields through the `index.max_docvalue_fields_search` setting. In a way, defaulting to doc values here bypasses that limit. I realize that runtime fields can already fetch through doc values, but maybe we have different performance/ consistency standards for calculated fields vs. runtime fields?

romseygeek replied:
To be consistent with how things are done for non-scripted values, I think we should remove the docvalues branch here.  It would be nice to be able to provide a hint at fetch time as to what the best way of loading values is (ie use docvalues if you're only loading a few values and aren't loading source, etc) but for now let's keep it simple.

javanna replied:
that means that we always recalculate, a bit like the fields API always fetches from _source?

romseygeek replied:
Yes. I think that's ok though, these are only used at fetch time.

javanna replied:
I agree

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,498 @@
+
+        @Override
+        public void document(int docID, StoredFieldVisitor visitor) throws IOException {
+            visitor.binaryField(SOURCE_FIELD_INFO, sourceBytes.toBytesRef().bytes);
```
Do we need to support other stored fields here, since we support `_fields` in runtime field scripts?

romseygeek replied:
We do! Will add that in.

romseygeek replied:
So this is possible, but it is not trivial as we need to do some careful handling of different types.  I'm now wondering if we should go the other way and remove stored fields access from runtime scripts - I know @jimczi was thinking that we should be discouraging stored fields use in general.

romseygeek replied:
I have implemented this properly with a test in MapperTestCase to ensure that it works for all mappers - turns out it wasn't quite as tricky as I thought, although it did involve a minor change to LeafStoredFieldsLookup to improve typing.

jtibshirani reviewed (commented):
I left a couple high-level thoughts as I got up to speed. The overall approach makes sense to me!

Do you plan to add some initial docs or open an associated issue? That could help make the API and behavior very clear during review. For example, I am curious how the `script` parameter will interact with others like `null_value` and `coerce`.

romseygeek commented:
> Do you plan to add some initial docs or open an associated issue? That could help make the API and behavior very clear during review. For example, I am curious how the script parameter will interact with others like null_value and coerce.

I will add some first-pass docs to this PR.  For `null_value` and `coerce`, I think we should not permit setting these if there is a `script` parameter on the mapper.  Mapper parameters can broadly be defined either as determining how something should be stored (`doc_values`, `index`, `stored`) or as describing some processing to be done on incoming values before storing (`null_value`, `coerce`, `ignore_malformed`).  The latter set should be handled within the script if there is one - for example, you can't emit null values on numeric field scripts, so you'd have to deal with that within the script anyway.

For error handling, I plan on adding a new field `on_script_error` which takes values of `ignore` or `reject`.  If you use `reject` then an exception thrown from the script will cause the whole document to be rejected.  `ignore` will store the name of the field in the doc's `_ignored` meta field and continue indexing.  We can also look at adding this to runtime scripts.  In both cases we will default to `reject`.

romseygeek committed:
Error handling

romseygeek committed:
Don't load from docvalues at fetch time

romseygeek committed:
tests

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
duh

romseygeek committed:
we don't sort emitted values

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Add parameter requirements and preclusions

romseygeek committed:
Add some basic docs

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Implement document() on lazy reader properly

romseygeek commented:
This is ready for another set of reviews:
* I've added error handling via a new parameter `on_script_error` that takes values `reject` (the default) or `ignore`.  This re-uses the existing `_ignored` metadata field for documents where an index-time script throws an error.  It's currently updateable, so you can change it from `reject` to `ignore` later if you decide that you want to continue indexing and fix things with a runtime field after you're discovered a bug.
* The document-backed  LeafReader now supports stored fields for all field types that can use it, and there is a new test in MapperTestCase that automatically checks this.
* I've added documentation for the `script` and `on_script_error` parameters to the number field mapper docs, and also restricted things so that you can't use `null_value`, `coerce` or `ignore_malformed` in combination with `script`.

I think that quite a few of the tests in CalculatedFieldTests will turn out to be generally useful for all mapping types that allows scripts, but I think it's better to leave them where they are for now and pull things into a base script testing class once we have more implementations.  I imagine we'll also want some more centralised docs for the general notion of index time scripts in addition to parameter definitions on individual field mapper doc pages but again I think that can wait for a follow up and will involve the docs team more closely.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -1088,6 +1228,32 @@ protected void parseCreateField(ParseContext context) throws IOException {
+                    context.searchLookup,
+                    context.leafReaderContext,
+                    0,
+                    v -> indexValue(context.pc, v));
```
if indexValue is called for every field value, don't we end up calling the _field_names code block once per value rather than once per field? That looks like it's how things already work without your change though.

romseygeek replied:
Yes, it's the same for both script and non-script code paths.  It would be nice to tidy this up in a followup - maybe we should store fields that need to be added to the field_names field on the ParseContext?

javanna replied:
ok it's unrelated, let's look at this again in a follow-up.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/PostParseExecutor.java:
```diff
@@ -0,0 +1,22 @@
+
+    void execute(PostParseContext context) throws IOException;
+
+    void onError(PostParseContext context, Exception e) throws IOException;
```
does the onError part change for each field type or could it be shared? I was assuming that it's the same, hence it does not even need to be made part of this interface, potentially?

romseygeek replied:
It needs to know about the `on_script_error` behaviour.  It might be simpler to merge that into the `execute` method though, I'll try it and see if it simplifies things.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+    // FieldExecutors can be called both by executePostParse() and from the lazy reader,
+    // so to ensure that we don't run field scripts multiple times we guard them with
+    // this one-time executor class
+    private class OneTimeFieldExecutor {
```
this is effectively to deal with scripts that mention other scripted fields, so that that same script is not executed again for the same document?

romseygeek replied:
Correct

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+                // this means that a mapper script is referring to another calculated field;
+                // in which case we need to execute that field first. We also check for loops here
+                if (fieldPath.add(field) == false) {
+                    throw new IllegalStateException("Loop in field resolution detected: " + String.join("->", fieldPath) + "->" + field);
```
I am not incredibly sure if this should be illegal state. It can effectively happen, so maybe it is more like an illegal argument exception?

romseygeek replied:
Oops, yes, that's me selecting the wrong exception type in autocomplete I think

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+                }
+                assert fieldExecutors.containsKey(field);
+                fieldExecutors.get(field).execute();
+                calculatedFields.remove(field);
```
I am missing why we remove from this set, is it so we execute the script only once?

romseygeek replied:
Yes - once it has been executed then its values are in the document and so will get picked up by the lazy reader normally, so we don't need to do this dance if it is referred to a second time.

javanna replied:
ok I think I get confused because I see executing the script once being a concern in two different places, one being the one time executor, and the other one being this set.

romseygeek replied:
Yes, it's not strictly necessary to remove it here but it means we can skip this whole branch once the field has been intercepted once.  I can remove it if you think it confuses things too much though.

javanna replied:
leave it for now, I will bump into it again in my next review and see

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+                    visitor.binaryField(fieldInfo, data);
+                }
+            }
+        }
```
is this effectively a fork of memory index, or the same code?

romseygeek replied:
It's completely new, MemoryIndex does not support stored fields.

javanna replied:
ok! could we add documentation about what all these bits are for, and what this has in common and not in common with memory index? I can see folks trying to refactor this to use memory index in the future without exactly knowing what's going on

javanna reviewed (commented):
Thanks for addressing my previous comments, I think that it is simpler than it was initially, but we may be able to simplify a little further, or maybe not :) I have left a bunch of comments and questions, hope this helps.

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -335,6 +370,26 @@ public Double parse(XContentParser parser, boolean coerce) throws IOException {
+                    public void executeAndEmit(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter) {
+                        DoubleFieldScript s = scriptFactory
+                            .newFactory(fieldName, script.getParams(), lookup)
+                            .newInstance(ctx);
```
The SearchLookup is unique per-document as well, so we need to rebuild the whole thing each time.  Fortunately this is a pretty simple object creation so it shouldn't have a significant impact on indexing time.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -335,6 +370,26 @@ public Double parse(XContentParser parser, boolean coerce) throws IOException {
+                    final DoubleFieldScript.Factory scriptFactory = service.compile(script, DoubleFieldScript.CONTEXT);
+
+                    @Override
+                    public void executeAndEmit(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter) {
```
We autobox in the indexing code already so I don't think trying to avoid it here is worth it, at least not until we've got some benchmarking runs in.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -958,6 +1059,28 @@ public ValueFetcher valueFetcher(SearchExecutionContext context, String format)
+                            script.executeAndEmit(context.lookup(), ctx, lookup.docId(), values::add);
+                        } catch (Exception e) {
+                            // ignore errors - if they exist here then they existed at index time
+                            // and were ignored, so we ignore here too
```
If `on_error` was set to `reject` then the document wouldn't have been indexed in the first place, so for it to get this far it must have been set to `ignore` when the doc was added.  So we ignore here as well.

javanna replied:
yea well, that makes sense. Maybe enforce the comment by saying "the only way the document could be indexed in the first time is that errors were ignored" .thanks for explaining

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+            false,
+            false,
+            IndexOptions.NONE,
+            DocValuesType.NONE,
```
The only thing that any of our field visitors actually use here is the name.  I'll add a comment explaining why we don't need to set anything else.

javanna replied:
thanks

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -823,6 +858,35 @@ protected void toXContent(XContentBuilder builder, boolean includeDefaults) thro
+                    Script script = Script.parse(o);
+                    if (script.getType() == ScriptType.STORED) {
+                        throw new IllegalArgumentException("stored scripts are not supported on scripted field [" + n + "]");
+                    }
```
I'm not sure I follow what you mean by 'the existing parse method'?

javanna replied:
sorry, I was referring to the AbstractScriptFieldType#parseScript which does exactly the same

romseygeek replied:
I think it's worth keeping them separate - it's only a few lines of code shared, and the parameter itself is sufficiently different (initializers and updateable values) that it's not worth the extra abstraction.

javanna replied:
ok, I was more worried about the handling to make sure that we don't support stored scripts, but indeed, the rest is just calling Script.parse so it is not a lot of code.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/PostParsePhase.java:
```diff
@@ -0,0 +1,533 @@
+     * @param lookup        the MappingLookup to collect executors from
+     * @param parseContext  the ParseContext of the current document
+     */
+    public static void executePostParsePhases(MappingLookup lookup, ParseContext parseContext) throws IOException {
```
I think I've put this method in about five different places, none of which feel quite right...  I originally had it on ParseContext but that felt wrong as context classes are really about holding state, not executing code.  And MappingLookup didn't feel right either, for the same reason.

javanna replied:
I understand :) though this way it really feels like "we could not find a better place for it, so we made it a public static method" :) I will see if I can come up with some other suggestions in the next review round

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -106,6 +108,8 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
+        PostParsePhase.executePostParsePhases(lookup, context);
+
         for (MetadataFieldMapper metadataMapper : metadataFieldsMappers) {
             metadataMapper.postParse(context);
```
So I originally had a `postParse` method on Mapper that was just called recursively here; however, this doesn't play well with scripted fields that refer to other scripted fields.  In order to make sure scripts are executed in the right order, we need to know up front which fields have scripts on them, so that in the special index-time lookup we can intercept a lookup on an unexecuted script field and execute it.  And we only want to execute a script once, so we need something to hold this state, which ends up giving us this PostParsePhase.  Once we're collecting this set of fields up-front anyway, it made sense to remove the general postParse method.  I agree about the naming though, I'll see if I can come up with something more descriptive.

javanna replied:
it is not only about naming though, it's the complexity added by the executor concept, and the collection of all the executors. Couldn't we have another way to collect upfront all of the fields that declare a script to achieve what the post parse phase achieves?

romseygeek committed:
feedback

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
DocumentLeafReader to top; execute functions to DocumentParser; rename to IndexTimeScript

romseygeek committed:
Remove MapperScript

romseygeek commented:
Latest reworking:
* PostParsePhase etc are gone, we now have a method on DocumentParser which makes much more sense
* LazyDocumentReader is now DocumentLeafReader (it's not particularly lazy anymore!) and lives in its own class
* PostParseExecutor is now IndexTimeScript
* MapperScript<T> is no longer a top-level class, although we still need an interface inside NumberFieldMapper so that we can build both long and double scripts.

romseygeek committed:
precommit

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java:
```diff
@@ -167,6 +174,14 @@ FieldTypeLookup fieldTypesLookup() {
+        return indexTimeLookup;
+    }
+
+    Map<String, IndexTimeScript> indexTimeScripts() {
```
could the index time script already have a reference to the index time lookup internally when it gets created, so that we don't need to expose this additional method in MappingLookup?

javanna replied:
having another look, index time lookup is needed when creating the search lookup, and I actually made another suggestion around possibly not needing IndexTimeScript entirely, so I don't think we will be able to do without adding indexTimeLooup.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -106,11 +115,61 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
         }
     }
 
+    private static void executeIndexTimeScripts(ParseContext context) {
```
I find this approach clearer than before, thanks! One thing I still notice is that a Script becomes a MapperScript, which becomes an IndexTimeScript which becomes a Consumer<LeafReaderContext>. Conceptually the same thing is represented in three different ways. I think I see the distinction between MapperScript and IndexTimeScript, as the former is executed against a real index to fetch values as well (could we do something to highlight that?). I am thinking that the OneTimeFieldExecutor could implement IndexTimeScript too, so that at least one step can be avoided. We do wrap the script but it is still something of the same class, which I find easier to reason about.

romseygeek replied:
I've folded OneTimeFieldExecutor directly into the consumer for DocumentLeafReader - is this any clearer?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -106,11 +115,61 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
+        }
+        SearchLookup searchLookup = new SearchLookup(
+            context.mappingLookup().indexTimeLookup()::get,
+            (ft, s) -> ft.fielddataBuilder(context.indexSettings().getIndex().getName(), s).build(
```
I find myself wondering what `s` is, as well as `k` and `c` or `pc` in other places. Can we use slightly longer variable names so it's easier to quickly know what these variable hold?

romseygeek replied:
Yeah this is a bad habit of mine, I'll make it clearer

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -120,6 +149,10 @@ public NumberFieldMapper build(ContentPath contentPath) {
     }
 
+    public interface MapperScript {
+        void executeAndEmit(SearchLookup lookup, LeafReaderContext ctx, int doc, Consumer<Number> emitter);
```
the emit terminology is already used for the emit function in scripts, and given that we have a consumer here, would it make sense to call this e.g. consumeValuesForDoc ?

romseygeek replied:
So I've moved this to a top-level class `FieldValues<T>` with a `valuesForDoc()` method.  It's a nice abstraction, actually, and I can see it being useful in other places as well later on.

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -734,6 +842,7 @@ public final NumericType numericType() {
         public final TypeParser parser() {
             return parser;
         }
+        public abstract BiFunction<Script, ScriptCompiler, MapperScript> compiler(String fieldName);
```
hopefully this can become mapperScript(String fieldName, Script script, ScriptCompiler compiler) ?

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/CalculatedFieldTests.java:
```diff
@@ -0,0 +1,338 @@
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
+
+public class CalculatedFieldTests extends MapperServiceTestCase {
```
given that we are not using the calculated terminology anywhere else at this point, maybe we should not use it here either. Do we need a specific test class for these tests or could they fit in the ordinary mapper tests? What tests will we add as we support more types?

romseygeek replied:
Some of these tests are generic 'does this work at all' tests, and some are more specific to the impls.  I think it will become clearer which are which when we add other field types.

javanna commented on server/src/test/java/org/elasticsearch/index/mapper/NumberFieldMapperTests.java:
```diff
@@ -226,5 +254,33 @@ protected final Object generateRandomInputValue(MappedFieldType ft) {
+            return (T) LongScriptFieldType.PARSE_FROM_SOURCE;
+        }
+        if (context == DoubleFieldScript.CONTEXT) {
+            return (T) DoubleScriptFieldType.PARSE_FROM_SOURCE;
```
what is the effect of using the parse from source script here?

romseygeek replied:
It's just a convenient already-built script that we can use to ensure that compilation etc are happening correctly.

javanna commented on x-pack/plugin/analytics/src/test/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapperTests.java:
```diff
@@ -49,6 +49,11 @@ protected void registerParameters(ParameterChecker checker) throws IOException {
     }
 
+    @Override
+    protected boolean supportsScripts() {
```
should we unify naming between this and `allowsIndexTimeScript` ?

romseygeek replied:
So this is not actually about index time scripts, it's about access to field data via a lookup.  Maybe it should be `supportsLookup` instead?

javanna commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:
```diff
@@ -106,11 +115,61 @@ private static void internalParseDocument(RootObjectMapper root, MetadataFieldMa
+
+    // field scripts can be called both by executeIndexTimeScripts() and via the document reader,
+    // so to ensure that we don't run field scripts multiple times we guard them with
+    // this one-time executor class
```
should this be javadocs?

javanna reviewed (commented):
I did another review round and left some more comments, I like the direction, it's getting simpler and hopefully we can simplify this even further, but it's close.

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentLeafReader.java:
```diff
@@ -0,0 +1,474 @@
+// does not attempt to do any analysis on text fields.  It also supports stored
+// fields where MemoryIndex does not.  It is used to back index-time scripts that
+// reference field data and stored fields from a document that has not yet been
+// indexed.
```
Because it was originally on a private class and I didn't convert it when I moved it to a top-level one.  Will update!

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -150,6 +183,13 @@ private float parseToFloat(Object value) {
+            @Override
+            public BiFunction<Script, ScriptCompiler, MapperScript> compiler(String fieldName) {
+                return (s, ss) -> {
+                    throw new IllegalArgumentException("Unknown parameter [script] for mapper [" + fieldName + "]");
```
It's a user error - we share an implementation for all the numeric types, so this can happen if somebody tries to add a mapping `{ "type" : "int", "script" : "blah" }`

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java:
```diff
@@ -168,6 +170,13 @@ public void parse(ParseContext context) throws IOException {
+    /**
+     * Returns an index time script for this field mapper, or {@code null} if none is defined
+     */
+    public IndexTimeScript getIndexTimeScript() {
```
Yeah, the tricky part here is that MapperScripts can be executed both at index time and at fetch time (for the value fetcher) and those two contexts have entirely different parameter requirements and so different signatures.  I think I can get rid of OneTimeFieldExecutor entirely and just have an anonymous consumer within DocumentParser, and I may be able to unify these two as well by having a class with two methods one it.

javanna replied:
yea I figured why MapperScript is there, and maybe that should be the only abstraction of a script for this functionality. I don't see the value of the indirection that IndexTimeScript causes now, as it could be just a method exposed directly by the mapper. Maybe you can give that a try and see what happens.

romseygeek replied:
We need to have some way of knowing whether or not a FieldMapper has a script on it, so that we can correctly intercept calls to those fields from the index-time SearchLookup and allow scripted fields to refer to each other.  The advantage of having this extra indirection is that it supplies this directly - if the IndexTimeScript is null, then the field doesn't have one.  If we move this to FieldMapper#executeScript then we need to add an additional `hasIndexTimeScript` boolean, which seems like the wrong trade-off to me?

javanna replied:
I would make FieldMapper script aware, more mappers will support it anyways, and FieldMapper already has getIndexTimeScript anyways for that purpose, plus executing the script itself. Would that work?

romseygeek replied:
OK, I have updated things:
* IndexTimeScript is gone
* FieldMapper has two methods, `hasIndexTimeScript` and `executeIndexTimeScript`

romseygeek commented on server/src/main/java/org/elasticsearch/index/mapper/NumberFieldMapper.java:
```diff
@@ -73,29 +80,43 @@ private static NumberFieldMapper toType(FieldMapper in) {
+            "on_script_error",
+            true,
+            m -> toType(m).onScriptError,
+            "reject", "ignore"
```
There are a few places we do this and we don't always have one.  Maybe if it becomes obvious that lots of stuff is being repeated when we add more impls?

romseygeek committed:
deef

romseygeek committed:
Die IndexTimeScript Die

romseygeek committed:
Merge remote-tracking branch 'origin/master' into runtime/runtime-to-index-3

romseygeek committed:
javadoc

jpountz commented on server/src/main/java/org/elasticsearch/index/mapper/DocumentLeafReader.java:
```diff
@@ -0,0 +1,476 @@
+            @Override
+            public int advance(int target) {
+                return 0;
+            }
```
We might expose ourselves to subtle bugs by not implementing the DocIdSetIterator contract correctly. Let's make sure that `docID()` returns -1 when the iterator is not positioned and make `nextDoc`/`advance` return `NO_MORE_DOCS` when appropriate?

javanna reviewed (approved):
thanks for all the iterations @romseygeek I left some super minor comments but this looks ready to me. The only thing that is left for us to settle on is the default behaviour when an error occurs running a script: reject or ignore? I know that @jimczi had an opinion on this. I am good with strict by default, and lenient when specifically configured. And maybe we should re-evaluate once we add the same handling to runtime fields.

romseygeek committed:
more

jtibshirani commented on docs/reference/mapping/types/numeric.asciidoc:
```diff
@@ -137,7 +139,26 @@ The following parameters are accepted by numeric types:
+
+`on_script_error`::
+
+    Defines what to do if the script defined by the `script` parameter
```
Sorry to be leaving a comment so late -- was just looking how this turned out! I was wondering if we considered re-using the `ignore_malformed` boolean flag here instead of introducing a new parameter? That feels simple and consistent to me, and avoids introducing a new enum parameter.

javanna replied:
thanks Julie, good question. It all started from "let's support some ignore_malformed behaviour" but then we decided to make it specific to scripts especially as we were not convinced that the two would have the same default values, as far as I recall. Another idea could be to support the new parameter as part of the definition of the script:

```
"field" : {
  "script" : {
    "source" : "here goes the script",
    "on_error" : "continue|fail"
  }
}
```

There is still time to change this if we want to, so let's discuss this.

romseygeek replied:
We did discuss re-using `ignore_malformed` but there are some important differences in the semantics of the two parameters.  In particular, `ignore_malformed` is specifically targeted at handling input data that cannot be converted into a number; because the scripts are typed, there is no way that they can return a malformed object, and so the same logic doesn't really apply.  In addition, a script could throw an error for any number of reasons (eg logic errors, array index out of bounds exceptions) and `ignore_malformed` feels too specific to cover all of these.

jpountz replied:
It's nice that we are reusing the `_ignored` meta field though. :+1:

jtibshirani replied:
Thanks everyone for the context. I agree the name would feel a bit off, maybe if it had been called `ignore_parse_errors` we could have reused it!

jpountz commented on docs/reference/mapping/types/numeric.asciidoc:
```diff
@@ -137,7 +139,26 @@ The following parameters are accepted by numeric types:
+`script`::
+
+    If this parameter is set, then the field will index values generated
+    by this script, rather than reading the values directly from the
```
The `rather` suggests that values that are set on the document would always be ignored, is it correct? I wonder if we should make it more explicit.

Do we have tests for this case (I couldn't find them from a quick look but I might have just missed them), as well as for the case when the script computes values based on content of the document under the _same_ field name?

javanna replied:
We reject documents that have a value for the field, if the field declares a script. We should probably clarify that in the docs.

romseygeek replied:
+1 to explicitly documenting this, I will open a followup

elasticmachine commented:
Pinging @elastic/es-search (Team:Search)
